### ISSUE TITLE
How to serialize Tianshou environment?

### ISSUE THREAD

**Issue by minhrongcon2000** [aai-institute/tianshou#496]
- [x] I have marked all applicable categories:
    + [ ] exception-raising bug
    + [ ] RL algorithm bug
    + [ ] documentation request (i.e. "X is missing from the documentation.")
    + [x] new feature request
- [x] I have visited the [source website](https://github.com/thu-ml/tianshou/)
- [x] I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates
- [x] I have mentioned version numbers, operating system and environment, where applicable:
  ```python
  import tianshou, torch, numpy, sys
  print(tianshou.__version__, torch.__version__, numpy.__version__, sys.version, sys.platform)
  ```
Hi,

I have an issue when using your application. I was trying to simulate the scenario of training agents in multiple client device, so I need to create multiple processes, each of which performs a custom training loop. 

When I use multiprocessing of python, I have encountered an issue of VecEnv not being able to be serialized. Do you guys have any solution to this?

Thank you and regards!

```python
print(tianshou.__version__, torch.__version__, numpy.__version__, sys.version, sys.platform)
0.4.4 1.9.0 1.20.3 3.8.8 (v3.8.8:024d8058b0, Feb 19 2021, 08:48:17) 
[Clang 6.0 (clang-600.0.57)] darwin
```

**Comment by Trinkle23897** [comment]
Maybe you should write `__getstate__` and `__setstate__` in your customized environment (or plus vector_env) so that pickle can successfully serialize the whole vector_env.
BTW, could you please provide a minimal example so that I can take a look at the issue and give one solution here?

**Comment by minhrongcon2000** [comment]
ok sure!

_TLDR:_ I attempted to use multiprocessing where each process is a client with their own OpenAI Gym environment. However, to boost the training process, I use your ``DummyVectorEnv``. When I parallel the process (see in the ``Server`` class), I got a bug of some components in class ``Client`` not being serializable. I have found out that lambda function and ``DummyVectorEnv`` are the only two components for the case. The former I can fixed; however, the latter depends entirely on the internal implementation of ``DummyVectorEnv``, so I'm afraid to touch this. I have tested on other vector env and have the same issues.

So here is an overview of my code!

This is some utilities class
```python
import gym
import tianshou as ts
import torch
import numpy as np
import random
import copy
from typing import Any, Callable, List
import multiprocessing as mp
import os


class Client:
    def __init__(self,
                 client_id: Any,
                 env: gym.Env,
                 policy: ts.policy.DQNPolicy,
                 batch_size: int,
                 epochs: int,
                 buffer: ts.data.ReplayBuffer,
                 exploration_update: Callable[[float], float],
                 step_per_collect: int,
                 test_num: int) -> None:
        
        self.client_id = client_id
        self.policy = policy
        
        self.buffer = buffer
        self.train_env = env
        self.test_env = env
        
        self.train_collector = ts.data.Collector(self.policy, 
                                                 self.train_env, 
                                                 self.buffer, 
                                                 exploration_noise=True)
        
        self.test_collector = ts.data.Collector(self.policy,
                                                self.test_env,
                                                exploration_noise=False)
        
        self.batch_size = batch_size
        self.epochs = epochs
        self.step_per_collect = step_per_collect
        self.test_num = test_num
        self.exp_update = exploration_update
        self.eps = 1
        self.mean_reward = None
        
    def update_weights(self):
        for _ in range(self.epochs):
            self.policy.set_eps(self.eps)
            self.train_collector.collect(n_step=self.step_per_collect)
            self.policy.update(self.batch_size, self.train_collector.buffer)
            
            test_result = self.test_collector.collect(n_episode=self.test_num)
            
            self.mean_reward = test_result["rew"].mean()
            
            self.eps = self.exp_update(self.eps)


class Server:
    def __init__(self,
                 policy: ts.policy.DQNPolicy,
                 chosen_prob: float) -> None:

        # global model info
        self.policy = policy
        
        # create clients
        self.client_pool: List[Client] = []
        self.chosen_prob = chosen_prob
        self.mean_reward = 0
        self.reward_std = 0
        
        self.num_round = 0
        
    def add_client(self, 
                   client_id: Any, 
                   env: gym.Env, 
                   local_batch_size: int, 
                   local_epochs: int, 
                   buffer: ts.data.ReplayBuffer,
                   exploration_update: Callable[[float], float],
                   step_per_collect: int,
                   test_num: int):
        
        new_client = DQNClient(
            client_id=client_id,
            env=env, 
            policy=copy.deepcopy(self.policy), 
            batch_size=local_batch_size,
            epochs=local_epochs,
            buffer=buffer,
            exploration_update=exploration_update,
            step_per_collect=step_per_collect,
            test_num=test_num
        )
        self.client_pool.append(new_client)
        
    def broadcast(self) -> None:
        for client in self.client_pool:
            client.policy.load_state_dict(copy.deepcopy(self.policy.state_dict()))

    def update_client(self, client):
        client.update_weights()
        self.logger.debug("Client {} finished update. Reward: {}".format(client.client_id, client.mean_reward))
        return client.mean_reward, copy.deepcopy(client.policy.state_dict())
        
    def aggregate(self) -> None:
        self.num_round += 1
        chosen_client: List[DQNClient] = random.sample(self.client_pool, 
                                                       int(len(self.client_pool) * self.chosen_prob))
        
        global_policy_weights = copy.deepcopy(self.policy.state_dict())
        
        rewards = []
        
        for key in global_policy_weights:
            global_policy_weights[key] = torch.zeros_like(global_policy_weights[key])
        
        # This is where serialize bug will happen
        # because the vector environment cannot be serialized and hence, no multiprocessing
        with mp.Pool(os.cpu_count()) as p:
            results = p.map(self.update_client, chosen_clients)
            for reward, weight in results:
                rewards.append(reward)
                for key in weight:
                    global_policy_weights[key] += 1 / len(chosen_client) * weight[key]
        
        self.policy.load_state_dict(global_policy_weights)
        self.mean_reward = np.mean(rewards)
        self.reward_std = np.std(rewards, ddof=1)
        
        self.logger.info("Round {} complete. Mean reward: {}, std reward: {}".format(self.num_round, self.mean_reward, self.reward_std))
        
    def save_model(self, model_dir):
        torch.save(self.policy, model_dir)
        
    def load_model(self, model_dir):
        self.policy.load_state_dict(torch.load(model_dir))
```

Here is a minimal training loop:

```python
set_global_seed()

# model setup
net: torch.nn.Module = ... 
optimizer = torch.optim.Adam(net.parameters(), lr=1e-3)
policy = ts.policy.DQNPolicy(net, optimizer, **kwargs) # some parameters are set in kwargs

# env setup
env = ts.env.DummyVectorEnv([lambda: gym.make("CartPole-v0") for _ in range(10)]

# server-client setup
server = Server(policy=policy, chosen_prob=0.1)
# # add 100 clients
for i in range(100): 
    server.add_client(
            client_id=i, 
            env=env, 
            local_batch_size=32, 
            local_epochs=10, 
            buffer=ts.data.VectorReplayBuffer(2000, 10), 
            exploration_update=lambda eps: max(eps * 0.99, 0.01), 
            step_per_collect=100, 
            test_num=10
        )


max_num_round = 200

if __name__ == "__main__:
    for _ in range(max_num_round):
        server.aggregate()
        server.broadcast()
        
        # some logging stuffs below
        ...
```

As I have commented in the code, in the ``aggregate`` method of class ``Server``, since class ``Client`` cannot be serialized due to encapsulating env and lambda function, the ``multiprocessing`` module cannot function properly. Sorry if my English and code quality are poor!

**Comment by Trinkle23897** [comment]
will take a look this weekend when I return to Pittsburgh, sorry about the late reply

**Comment by minhrongcon2000** [comment]
Oh sorry! I forgot to reply! I just fixed that bug today. It turns out this bug is due to out-dated version. This can totally be resolved by updating to the latest one. You can close the issue now! Thanks for your help anw!
